src/rhsmlib/cloud/_base_provider.py

```diff
@@ -81,16 +81,90 @@ class attributes. Logic of gathering metadata/signature will be implemented
     # When a token is supported by cloud provider, then this value is in seconds
     CLOUD_PROVIDER_TOKEN_TTL = None
 
-    def __init__(self) -> None:
+    # Time to live of in-memory cache for metadata and signature (value is in seconds)
+    IN_MEMORY_CACHE_TTL = 10.0
+
+    def __init__(self, hw_info: dict = None):
         """
-        Initialize instance of CloudCollector.
+        Initialize cloud provider
+        :param hw_info: Dictionary with hardware information.
         """
+        # In-memory cache of metadata
+        self._cached_metadata: str = None
+        # Time, when metadata was received. The value is in seconds (unix time)
+        self._cached_metadata_ctime: float = None
+        # In-memory cache of signature
+        self._cached_signature: str = None
+        # Time, when signature was received. The value is in seconds (unix time)
+        self._cached_signature_ctime: float = None
+
+        # Dictionary with hardware information
+        if hw_info is None:
+            self.hw_info: dict = self.collect_hw_facts()
+        else:
+            self.hw_info: dict = hw_info
+
         # In-memory cache of token. The token is simple string
-        self._token = None
+        self._token: str = None
         # Time, when token was received. The value is in seconds (unix time)
-        self._token_ctime = None
+        self._token_ctime: float = None
         # Time to Live of token
-        self._token_ttl = None
+        self._token_ttl: float = None
+
+    def __lt__(self, other) -> bool:
+        """
+        All instances of BaseCloudProvider are equal
+        :param other: instance of other cloud provider
+        :return: False
+        """
+        return False
+
+    def __gt__(self, other) -> bool:
+        """
+        All instances of BaseCloudProvider are equal
+        :param other: instance of other cloud provider
+        :return: False
+        """
+        return False
+
+    def __eq__(self, other) -> bool:
+        """
+        All instances of BaseCloudProvider are equal
+        :param other: instance of other cloud provider
+        :return: False
+        """
+        return True
```

Comment:
if there is, actually, no ordering possible between providers, does it still make sense to implement these? what does happen if they are not provided?

Reply:
We sort cloud providers using priorities. When two providers have same priority and there there are no methods for comparing instances of `BaseCloudProvider` or its subclass, then sorting fails. I was considered adding priority as property of `BaseCloudProvider`, but I don't like properties with complicated side effects (calling subprocessess).

Reply:
since these classes are all internal, wouldn't be easier to fix the sorting in `_get_cloud_providers()` to consider only the priority? ie
```python
cloud_list.sort(key=lambda x: x[0], reverse=True)
```
This way we can avoid these "not exactly useful" implementations (which could be silently misused in other contexts).

Reply:
Yeah, I like the trick with the `key`.  :+1: